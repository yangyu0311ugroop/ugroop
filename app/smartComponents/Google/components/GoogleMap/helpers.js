import { WAYPOINTS_LIMIT } from 'appConstants';
import first from 'lodash/first';

const normaliseWaypoint = waypoints => (accu, order) => {
  const {
    location: { placeId },
  } = waypoints[order];
  return accu.concat(placeId);
};

const normaliseData = direction => {
  const { request, routes } = direction;
  const {
    bounds,
    legs,
    waypoint_order: waypointOrders,
    summary,
    overview_polyline: polyline,
  } = routes[0];
  const {
    origin: { placeId: origin },
    destination: { placeId: destination },
    waypoints,
  } = request;

  const normaliseWaypoints = waypointOrders.reduce(
    GOOGLE_API_HELPERS.normaliseWaypoint(waypoints),
    [],
  );

  return {
    bounds,
    legs,
    route: [origin, ...normaliseWaypoints, destination],
    summary,
    polyline,
  };
};

const normaliseLeg = (
  {
    [locationKey || 'start_location']: location,
    distance: { value: legDistance },
    duration: { value: legDuration },
  },
  marker,
  locationKey,
) => ({
  marker: {
    lat: location.lat(),
    lng: location.lng(),
    ...marker,
  },
  distance: legDistance,
  duration: legDuration,
});

const normaliseLegs = legs => {
  const {
    start_location: start,
    end_location: end,
    distance: { text: legDistance },
    duration: { text: legDuration },
  } = first(legs);

  return {
    startMarker: {
      lat: start.lat(),
      lng: start.lng(),
    },
    endMarker: {
      lat: end.lat(),
      lng: end.lng(),
    },
    distance: legDistance,
    duration: legDuration,
  };
};

const normaliseMarker = ({
  index,
  legs,
  routeIds,
  isLast,
  route,
  ids = [],
}) => ({ nodes, markers, distance, duration }, placeId, idx) => {
  const isFirst = index === 0;

  // legs length smaller than route length by 1
  if (idx > legs.length - 1) return { nodes, markers, distance, duration };

  const {
    start_location: startLocation,
    distance: { value: legDistance },
    duration: { value: legDuration },
  } = legs[idx];

  const startId = routeIds[index * WAYPOINTS_LIMIT + idx];
  const start = {
    id: startId,
    index: idx,
    placeId,
    origin: isFirst && idx === 0,
    destination: isLast && idx === route.length - 1,
    lat: startLocation.lat(),
    lng: startLocation.lng(),
    distance: legDistance,
    duration: legDuration,
  };
  const nodeStart = {
    [startId]: {
      calculated: {
        index: ids.indexOf(startId),
        distance: legDistance,
        origin: isFirst && idx === 0,
        destination: isLast && idx === route.length - 1,
      },
    },
  };

  // process last leg
  if (isLast && idx === legs.length - 1) {
    const { end_location: endLocation } = legs[idx];

    const endPlaceId = route[idx + 1];
    const endId = routeIds[index * WAYPOINTS_LIMIT + idx + 1];
    const end = {
      id: endId,
      index: idx + 1,
      placeId: endPlaceId,
      origin: isFirst && idx === 0,
      destination: isLast && idx === route.length - 2,
      lat: endLocation.lat(),
      lng: endLocation.lng(),
    };
    const nodeEnd = {
      [endId]: {
        calculated: {
          index: ids.indexOf(endId),
          origin: isFirst && idx === 0,
          destination: isLast && idx === route.length - 2,
        },
      },
    };

    return {
      markers: markers.concat([start, end]),
      nodes: { ...nodes, ...nodeStart, ...nodeEnd },
      distance: distance + legDistance,
      duration: duration + legDuration,
    };
  }

  // if the route is not the first part, ignore the first one
  // if (index > 0 && idx === 0) return { markers, distance, duration };

  // process other leg
  return {
    markers: markers.concat([start]),
    nodes: { ...nodes, ...nodeStart },
    distance: distance + legDistance,
    duration: duration + legDuration,
  };
};

const normaliseDirection = ({ index, direction, isLast }, props) => {
  const {
    bounds,
    legs,
    route,
    summary,
    polyline,
  } = GOOGLE_API_HELPERS.normaliseData(direction);

  return {
    bounds,
    summary,
    polyline,
    ...route.reduce(
      GOOGLE_API_HELPERS.normaliseMarker({
        index,
        legs,
        isLast,
        route,
        ...props,
      }),
      {
        nodes: {},
        markers: [],
        distance: 0,
        duration: 0,
      },
    ),
  };
};

export const GOOGLE_API_HELPERS = {
  normaliseWaypoint,
  normaliseData,
  normaliseLeg,
  normaliseLegs,
  normaliseMarker,
  normaliseDirection,
};
