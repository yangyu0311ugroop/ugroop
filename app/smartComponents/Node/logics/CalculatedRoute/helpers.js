import dotProp from 'dot-prop-immutable';

let fetchedRoutes = {
  test: 'test',
  testMerge: { id: 2233 },
};

const getCache = (prop, value) =>
  prop && dotProp.get(fetchedRoutes, prop, value);
const setCache = (prop, value) => {
  fetchedRoutes = dotProp.set(fetchedRoutes, `${prop}`, value);
};
const mergeCache = (prop, value) => {
  fetchedRoutes = dotProp.merge(fetchedRoutes, prop, value);
};
const deleteCache = routeId => {
  fetchedRoutes = dotProp.delete(fetchedRoutes, `${routeId}`);
};
const isValid = raw => {
  if (!Array.isArray(raw) || !raw.length) {
    return false;
  }

  for (let i = 0; i < raw.length; i += 1) {
    if (!raw[i]) return false;
  }

  return true;
};

export const ROUTE_CACHE = {
  fetchedRoutes,
  delete: deleteCache,
  get: getCache,
  set: setCache,
  merge: mergeCache,
  isValid,
};
