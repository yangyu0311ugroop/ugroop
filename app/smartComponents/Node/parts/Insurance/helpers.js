import { INSURANCE_MODES } from 'smartComponents/Node/parts/Insurance/constants';

export const isDateMode = mode => {
  switch (mode) {
    case INSURANCE_MODES.pending:
    case INSURANCE_MODES.confirmed:
      return true;

    default:
      return false;
  }
};

export const renderMode = mode => {
  switch (mode) {
    case INSURANCE_MODES.pending:
      return 'Application sent';
    case INSURANCE_MODES.confirmed:
      return 'Confirmation received';
    case INSURANCE_MODES.other:
      return 'Other';
    default:
      return null;
  }
};

export const INSURANCE_HELPERS = {
  isDateMode,
  renderMode,
};
