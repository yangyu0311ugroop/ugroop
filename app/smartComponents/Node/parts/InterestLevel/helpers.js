import React from 'react';
import { FormattedMessage as M } from 'react-intl';
import { NOT_APPLICABLE } from 'appConstants';
import { INTEREST_LEVELS } from './constants';
import m from './messages';

const getMessage = value => m[value];

const renderValue = value => {
  const message = getMessage(value);
  if (message) return <M {...message} />;
  return null;
};

const makeOptions = required => {
  const interestLevels = Object.values(INTEREST_LEVELS);

  if (!required) {
    interestLevels.unshift(NOT_APPLICABLE);
  }

  return interestLevels.map(value => ({
    value,
    children: renderValue(value),
  }));
};

export default {
  getMessage,
  renderValue,
  makeOptions,
};
