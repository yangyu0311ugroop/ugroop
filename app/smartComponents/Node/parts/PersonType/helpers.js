import React from 'react';
import { FormattedMessage as M } from 'react-intl';
import { PERSON_TYPES } from 'smartComponents/Node/parts/PersonType/constants';
import m from './messages';

const getMessage = value => m[value];

const renderValue = value => {
  const values = Object.keys(m);
  const includesValue = values.includes(value);
  if (includesValue) {
    const message = getMessage(value);
    if (message) return <M {...message} />;
  }

  return 'Other';
};

const renderOtherValue = value => value;

const makeOptions = () =>
  Object.values(PERSON_TYPES).map(value => ({
    value,
    children: renderValue(value),
  }));

export default {
  getMessage,
  renderValue,
  makeOptions,
  renderOtherValue,
};
