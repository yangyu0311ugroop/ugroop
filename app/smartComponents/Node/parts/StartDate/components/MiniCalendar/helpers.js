import moment from 'moment';

export const WEEKDAYS = [0, 1, 2, 3, 4, 5, 6];

// react-dates helper function
export default function getCalendarMonthWeeks(
  startDate,
  duration,
  firstDayOfWeek = moment.localeData().firstDayOfWeek(),
) {
  const month = moment(startDate);
  const endMonth = moment(startDate).add(duration, 'day');

  if (!moment.isMoment(month) || !month.isValid()) {
    return null;
  }
  if (!moment.isMoment(endMonth) || !endMonth.isValid()) {
    return null;
  }
  if (WEEKDAYS.indexOf(firstDayOfWeek) === -1) {
    return null;
  }

  // set utc offset to get correct dates in future (when timezone changes)
  const firstOfMonth = month
    .clone()
    .startOf('month')
    .hour(12);
  const lastOfMonth = endMonth
    .clone()
    .endOf('month')
    .hour(12);

  // calculate the exact first and last days to fill the entire matrix
  // (considering days outside month)
  const prevDays = (firstOfMonth.day() + 7 - firstDayOfWeek) % 7;
  const nextDays = (firstDayOfWeek + 6 - lastOfMonth.day()) % 7;
  const firstDay = firstOfMonth.clone().subtract(prevDays, 'day');
  const lastDay = lastOfMonth.clone().add(nextDays, 'day');

  const totalDays = lastDay.diff(firstDay, 'days') + 1;

  const currentDay = firstDay.clone();
  const weeksInMonth = [];

  for (let i = 0; i < totalDays; i += 1) {
    if (i % 7 === 0) {
      weeksInMonth.push([]);
    }

    let day = null;
    if (i >= prevDays && i < totalDays - nextDays) {
      day = currentDay.clone();
    }

    weeksInMonth[weeksInMonth.length - 1].push(day);

    currentDay.add(1, 'day');
  }

  return weeksInMonth;
}
