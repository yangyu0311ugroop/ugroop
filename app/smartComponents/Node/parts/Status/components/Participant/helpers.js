import React from 'react';
import { FormattedMessage as M } from 'react-intl';
import {
  PARTICIPANT_STATUSES,
  PARTICIPANT_STATUSES_REQUIRED,
} from 'utils/constants/nodes';
import m from './messages';

const getMessage = value => m[value];

const renderValue = value => {
  const message = getMessage(value);
  if (message) return <M {...message} />;
  return null;
};

const makeOptions = required => {
  const participantStatuses = required
    ? PARTICIPANT_STATUSES_REQUIRED
    : PARTICIPANT_STATUSES;
  const participantStatusArray = Object.values(participantStatuses);

  return participantStatusArray.map(value => ({
    value,
    children: renderValue(value),
  }));
};

export default {
  getMessage,
  renderValue,
  makeOptions,
};
