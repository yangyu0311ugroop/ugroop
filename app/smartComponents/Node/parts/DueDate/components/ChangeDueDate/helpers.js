import { DEFAULT, HUMANIZE_DURATION_CONSTANT } from 'appConstants';
import momentjs from 'moment/moment';
import React, { Fragment } from 'react';
import { FORMATS_DATE_TIME } from 'utils/constants/dateTime';
import { LOGIC_HELPERS } from 'utils/helpers/logic';
import moment from 'utils/helpers/moment';
import { DAY, TEMPLATE, ACTIVITY, CHECKGROUP } from 'utils/modelConstants';
import humanizeDuration from 'humanize-duration';

// temporary treat activity as event for testing
const EVENT = ACTIVITY;

export const RELATIVE = 'relative';
export const FIXED = 'fixedDateTime';
export const FIXED_NO_TIME = 'fixedDate';
export const UNSET = 'unset';
export const CUSTOM = 'custom';
export const DEFAULT_DURATION = 'P0D';
export const DEFAULT_DURATION_TIME = 'PT0S';

export const DUE_DATE_RELATIVE_OPTIONS = {
  [TEMPLATE]: ['-P1D', '-P2D', '-P7D', '-P14D', '-P1M', '-P2M'],
  [DAY]: ['-P1D', '-P2D', '-P7D'],
  [EVENT]: ['-PT5M', '-PT15M', '-PT30M', '-PT1H'],
  [CHECKGROUP]: [],
  [DEFAULT]: [],
};

const isUnset = mode => mode === UNSET;
const isCustom = mode => mode === CUSTOM;
const isFixed = mode => mode === FIXED || mode === FIXED_NO_TIME;
const isFixedWithTime = mode => mode === FIXED;
const isFixedNoTime = mode => mode === FIXED_NO_TIME;
const isRelative = mode => mode === RELATIVE;
const isDefault = value => value === DEFAULT_DURATION;
const isDefaultTime = value => value === DEFAULT_DURATION_TIME;
const inADay = dayCount => dayCount < 1 && dayCount > 0;
const aDayAgo = dayCount => dayCount > -1 && dayCount < 0;
const withinADay = dayCount => dayCount < 1 && dayCount > -1;
const isOneDayMinimum = parentType =>
  !parentType || parentType === TEMPLATE || parentType === DAY;

const makeRelativeOption = (key, value) => {
  if (!key) {
    return null;
  }

  if (value) {
    return { mode: RELATIVE, key, value };
  }

  return DUE_DATE_HELPERS.makeCustomDurationOption(key);
};

const makeRelativeOptions = keys =>
  Array.isArray(keys)
    ? keys.map(key => DUE_DATE_HELPERS.makeRelativeOption(key))
    : [];

const renderNoOffset = parentType => {
  switch (parentType) {
    case TEMPLATE:
      return 'start of tour';
    case DAY:
      return 'on this day';
    case EVENT:
      return 'on date of event';
    default:
      return 'start of tour, day or event';
  }
};

const modeEventOptions = () => [
  DUE_DATE_HELPERS.makeRelativeOption(
    'PT0S',
    <Fragment>at time of event</Fragment>,
  ),
  ...DUE_DATE_HELPERS.makeRelativeOptions(DUE_DATE_RELATIVE_OPTIONS[EVENT]),
];

const makeOptionsByType = parentType => {
  switch (parentType) {
    case TEMPLATE:
    case DAY:
      return DUE_DATE_HELPERS.makeRelativeOptions(
        DUE_DATE_RELATIVE_OPTIONS[parentType],
      );
    case EVENT:
      return DUE_DATE_HELPERS.modeEventOptions();
    default:
      return DUE_DATE_HELPERS.makeRelativeOptions(
        DUE_DATE_RELATIVE_OPTIONS[DEFAULT],
      );
  }
};

const makeCustomDurationOption = (
  value,
  parentType,
  useActual = false,
  prefix,
) => {
  if (!value) {
    return null;
  }

  const humaniseDuration = DUE_DATE_HELPERS.humaniseDuration({
    value,
    parentType,
    useActual,
  });

  return {
    mode: RELATIVE,
    key: value,
    value: LOGIC_HELPERS.ifElse(
      prefix,
      <span>
        Click to set {'"'}
        {humaniseDuration}
        {'"'}
      </span>,
      humaniseDuration,
    ),
  };
};

const makeSelectedFixedDateOption = dueDate => {
  if (!dueDate) {
    return null;
  }
  const { mode, value } = dueDate;

  if (!DUE_DATE_HELPERS.isFixed(mode)) {
    return null;
  }

  const key = momentjs(value).toISOString();

  if (DUE_DATE_HELPERS.isFixedWithTime(mode)) {
    return {
      mode,
      key,
      value: (
        <Fragment>
          {moment.renderDayDate(value)} at {moment.renderTime(value)}
        </Fragment>
      ),
    };
  }

  return {
    mode,
    key,
    value: <Fragment>{moment.renderDayDate(value)}</Fragment>,
  };
};

const makeSelectedRelativeDateOption = (
  dueDate,
  parentType,
  useActual = false,
) => {
  if (!dueDate) {
    return null;
  }

  const { mode, value } = dueDate;

  if (
    DUE_DATE_HELPERS.isDefault(value) ||
    DUE_DATE_HELPERS.isDefaultTime(value)
  ) {
    return null;
  }

  if (!DUE_DATE_HELPERS.isRelative(mode)) {
    return null;
  }

  return DUE_DATE_HELPERS.makeCustomDurationOption(
    dueDate.value,
    parentType,
    useActual,
  );
};

const makeSelectedFixedDateOptionState = state => {
  const { fixedDate, fixedTime } = state;

  if (!fixedDate) {
    return null;
  }
  const dateTime = moment.setTime(
    momentjs(fixedDate),
    momentjs(fixedTime, FORMATS_DATE_TIME.TIME),
  );
  const mode = LOGIC_HELPERS.ifElse(fixedTime, FIXED, FIXED_NO_TIME);

  return {
    mode,
    key: `${dateTime.toISOString()}`,
    value: (
      <Fragment>
        Click to set {`"`}
        {moment.renderDayDate(dateTime)}
        {LOGIC_HELPERS.ifElse(fixedTime, ` at ${moment.renderTime(dateTime)}`)}
        {`"`}
      </Fragment>
    ),
  };
};

const makeNoOffsetOption = parentType => ({
  mode: RELATIVE,
  key: DEFAULT_DURATION,
  value: <Fragment>{DUE_DATE_HELPERS.renderNoOffset(parentType)}</Fragment>,
});

const makeHeadingOption = value => ({
  key: value,
  value,
  heading: true,
  disabled: true,
});

const renderOffset = ({ value, parentType, translateType, useActual }) => {
  const duration = momentjs.duration(value);
  const durationInMilliSec = duration.asMilliseconds();

  const offsetDirection = LOGIC_HELPERS.ifElse(
    value && value.indexOf('-') === -1,
    'after',
    'before',
  );
  const postfix = LOGIC_HELPERS.ifElse(
    [translateType, parentType !== CHECKGROUP],
    ` this ${translateType || ''}`,
    '',
  );
  let renderDuration;
  let durationSet = duration.humanize();

  if (useActual) {
    durationSet = humanizeDuration(
      durationInMilliSec,
      HUMANIZE_DURATION_CONSTANT,
    );
  }

  switch (parentType) {
    case TEMPLATE:
    case DAY:
      renderDuration = LOGIC_HELPERS.ifElse(
        DUE_DATE_HELPERS.withinADay(duration.asDays()),
        'a day',
        durationSet,
      );
      break;
    default:
      renderDuration = durationSet;
      break;
  }

  return `${renderDuration} ${offsetDirection}${postfix}`;
};

const humaniseDuration = ({ value, parentType, ...props }) => {
  if (!value) {
    return null;
  }

  if (DUE_DATE_HELPERS.isDefault(value)) {
    return DUE_DATE_HELPERS.renderNoOffset(parentType);
  }
  if (DUE_DATE_HELPERS.isDefaultTime(value)) {
    return 'at time of event';
  }

  return DUE_DATE_HELPERS.renderOffset({ value, parentType, ...props });
};

const calculateDate = ({ startTime, dueDate, parentType, anchorDate }) => {
  if (startTime) return startTime;

  if (!dueDate) {
    return null;
  }
  const { mode, value } = dueDate;

  if (DUE_DATE_HELPERS.isFixed(mode)) {
    return momentjs(value);
  }

  if (!DUE_DATE_HELPERS.isRelative(mode)) {
    return null;
  }

  const duration = momentjs.duration(value);

  if (!anchorDate) {
    return null;
  }

  const dayCount = duration.asDays();

  if (!DUE_DATE_HELPERS.isOneDayMinimum(parentType)) {
    return momentjs(anchorDate).add(duration);
  }
  if (DUE_DATE_HELPERS.inADay(dayCount)) {
    return momentjs(anchorDate).add(1, 'day');
  }
  if (DUE_DATE_HELPERS.aDayAgo(dayCount)) {
    return momentjs(anchorDate).add(-1, 'day');
  }

  return momentjs(anchorDate).add(duration);
};

const renderCalculatedDate = ({ value, parentType, calculatedDate }) => {
  const duration = momentjs.duration(value);
  const dayCount = duration.asDays();

  if (
    !DUE_DATE_HELPERS.isDefault(value) &&
    !DUE_DATE_HELPERS.isOneDayMinimum(parentType) &&
    DUE_DATE_HELPERS.withinADay(dayCount)
  ) {
    return moment.renderTime(calculatedDate);
  }

  return moment.renderCalendarDueDate(calculatedDate);
};

export const DUE_DATE_HELPERS = {
  isUnset,
  isCustom,
  isFixed,
  isFixedWithTime,
  isFixedNoTime,
  isRelative,
  isDefault,
  isDefaultTime,
  inADay,
  aDayAgo,
  withinADay,
  isOneDayMinimum,
  modeEventOptions,
  makeRelativeOption,
  makeRelativeOptions,
  makeSelectedFixedDateOption,
  makeSelectedFixedDateOptionState,
  makeOptionsByType,
  makeNoOffsetOption,
  makeHeadingOption,
  renderOffset,
  renderNoOffset,
  makeCustomDurationOption,
  makeSelectedRelativeDateOption,
  humaniseDuration,
  calculateDate,
  renderCalculatedDate,
};
