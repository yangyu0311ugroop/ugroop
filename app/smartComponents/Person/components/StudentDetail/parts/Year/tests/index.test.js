import React from 'react';
import { shallow } from 'enzyme';
import { TEST_HELPERS } from 'utils/helpers/testHelpers';
import { StudentDetailYear } from '../index';

describe('<StudentDetailYear />', () => {
  let rendered;
  let instance;

  beforeEach(() => {
    rendered = shallow(<StudentDetailYear />);
    instance = rendered.instance();
  });

  describe('renderTextOnly', () => {
    it('should match snapshot', () => {
      TEST_HELPERS.expectMatchSnapshot(instance.renderTextOnly);
    });
  });

  describe('renderTextField', () => {
    it('should match snapshot', () => {
      TEST_HELPERS.expectMatchSnapshot(instance.renderTextField);
    });
  });
});
