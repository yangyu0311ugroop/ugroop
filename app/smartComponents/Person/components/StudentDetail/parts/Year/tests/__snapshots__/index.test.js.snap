// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StudentDetailYear /> renderTextField should match snapshot 1`] = `
<div>
  <TextField
    autoFocus={true}
    label={
      <FormattedMessage
        defaultMessage="Year Level"
        id="app.smartComponents.Person.StudentDetail.Year.label"
      />
    }
    maxLength={0}
    minLength={0}
    name="studentDetail.year"
    type="text"
    validationErrors={Object {}}
    validations={Object {}}
    value=""
  />
</div>
`;

exports[`<StudentDetailYear /> renderTextOnly should match snapshot 1`] = `
<div>
  <Span
    dense={true}
    transform="capitalize"
  />
</div>
`;
